### PR TITLE
Fix/fetch product concurrency

### LINKED ISSUE #344: Crash While Requesting Product Info
### Platform

- [x] iOS
- [ ] macOS
- [ ] tvOS

### In app purchase type

- [ ] Consumable
- [ ] Non-consumable 
- [x] Auto-Renewable Subscription
- [ ] Non-Renewing Subscription

### Environment

- [ ] Sandbox
- [x] Production

### Version
ℹ Please replace this with the version of SwiftyStoreKit you're using.

### Issue summary
ℹ While requesting product info some time app is suddenly crashed.  

### What did you expect to happen
ℹ App should not crash

### What happened instead
ℹ App is crashed occasionally.

Following is our code Snippet.
func requestProductData(productIds: Set<String>, productOrderedList:NSMutableArray,  viewController:UIViewController?) {
        
        orderedList = productOrderedList
        DLog("orderedList \(orderedList)")
        DLog("product list \(productIds)")
        DLog("productOrderedList \(productOrderedList)")
        
        currentController = viewController
   
        SwiftyStoreKit.retrieveProductsInfo(productIds) { result in
            self.productList = Array(result.retrievedProducts)
            self.displayProductInfo()
            
        }
    }

We are calling above Function From Background Queue.Following is crash SS.
screen shot 2018-01-27 at 11 57 15 am

### LINKED ISSUE #468: Crash in ProductsInfoController.retrieveProductsInfo
### Platform

- [x] iOS
- [ ] macOS
- [ ] tvOS

### In app purchase type

- [ ] Consumable
- [x] Non-consumable 
- [ ] Auto-Renewable Subscription
- [ ] Non-Renewing Subscription

### Environment

- [x] Sandbox
- [x] Production

### Version
0.15.0

### Issue summary

sometimes, it crash when it retrieve products info. 

```0  SwiftyStoreKit                 0x1014a6db4 closure #1 in ProductsInfoController.retrieveProductsInfo(_:completion:) (<compiler-generated>)
1  SwiftyStoreKit                 0x101497258 partial apply for closure #1 in InAppProductQueryRequest.performCallback(_:) (InAppProductQueryRequest.swift:76)
    2  SwiftyStoreKit                 0x1014a0f34 thunk for @escaping @callee_guaranteed () -> () (<compiler-generated>)
3  libdispatch.dylib              0x2320bca38 _dispatch_call_block_and_release + 24
4  libdispatch.dylib              0x2320bd7d4 _dispatch_client_callout + 16
5  libdispatch.dylib              0x23209d9e4 _dispatch_main_queue_callback_4CF$VARIANT$armv81 + 1008
6  CoreFoundation                 0x23260dec0 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
7  CoreFoundation                 0x232608df8 __CFRunLoopRun + 1924
8  CoreFoundation                 0x232608354 CFRunLoopRunSpecific + 436
9  GraphicsServices               0x23480879c GSEventRunModal + 104
10 UIKitCore                      0x25ebf3b68 UIApplicationMain + 212
11 app.                               0x100438400 main (main.m:14)
12 libdyld.dylib                  0x2320ce8e0 start + 4```

### PR DESCRIPTION
**Pull request by cipolleschi**
This PR Fixes #344 (outside of the main thread) and fixes #468.

### CODE REVIEW

**Comment by bizz84** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please target PRs to <code>develop</code> branch</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">No CHANGELOG changes made</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by cipolleschi** [comment]
updated.

**Comment by cipolleschi** [comment]
I have some trouble making the test work with the CI.
Locally, they work perfectly: the execution is paused, waiting for the thread to signal the completion of the chunk of work and verify the expectation afterward.
Anyway, once executed by Travis, it looks like that the execution is not paused, and so the test fails.

Do you know how to fix this?

File: SwiftyStoreKit/ProductsInfoController.swift [hunk 1]
```diff
@@ -44,18 +44,29 @@ class ProductsInfoController: NSObject {
     }
     
     let inAppProductRequestBuilder: InAppProductRequestBuilder
+
+    private var spinLock: OSSpinLock
```
Comment (Sam-Spencer) [inline SwiftyStoreKit/ProductsInfoController.swift]: I'm under the impression that `OSSpinLock` is deprecated in favor of `os_unfair_lock`. Is there a reason you use this here but not in tests and vice versa?

**Comment by Sam-Spencer** [comment]
Closing PR for the time being as *stale*.

**Comment by RamyaThirunavukkarasu** [comment]
I am getting crashes while fetching the promotional code discount price details (localizedPrice and  priceLocale.currencyCode) in iOS 12.5 and 13.5 but iOS 14 is working fine .

static func getProductOfferDetails(_ productDiscount: SKProductDiscount?) -> InAppProductDiscount? {
        guard let productDiscount = productDiscount else { return nil }
        
        var inAppProductDiscunt = InAppProductDiscount()
        inAppProductDiscunt.price = productDiscount.localizedPrice ?? String()
        inAppProductDiscunt.currency_code = productDiscount.priceLocale.currencyCode
        inAppProductDiscunt.numberOfPeriods = productDiscount.numberOfPeriods
        inAppProductDiscunt.numberOfUnits = productDiscount.subscriptionPeriod.numberOfUnits
        inAppProductDiscunt.localizedSubscriptionPeriod = productDiscount.localizedSubscriptionPeriod
        inAppProductDiscunt.unit = InAppProduct.getSubscriptionPeriodUnit(productDiscount.subscriptionPeriod.unit.rawValue)
        inAppProductDiscunt.paymentMode = inAppProductDiscunt.getPaymentMode(productDiscount.paymentMode.rawValue)
        if #available(iOS 12.2, *) {
            inAppProductDiscunt.type = inAppProductDiscunt.getType(productDiscount.type.rawValue)
            inAppProductDiscunt.identifier = productDiscount.identifier
        }
        return inAppProductDiscunt
    }

**Pull request merged**